src/matching/MatchingReducerFactory.js

```diff
@@ -0,0 +1,100 @@
+import { matcher as defaultMatcher } from './matchers';
+import { warn } from '../utils/logger';
+
+const identity = value => value;
+
+/**
+ * @class MatchingReducerFactory
+ *
+ * A simple pattern matching abstraction
+ * for redux reducers.
+ */
+export default class MatchingReducerFactory {
+
+  /**
+   * @param {any} Initial app state
+   * @param {Matcher} Default Matcher implementation
+   */
+  constructor(initialAppState, defaultMatcherImpl = defaultMatcher) {
+    this.defaultMatcherImpl = defaultMatcherImpl;
+    this.initialAppState = initialAppState;
+    this.matchersWithHandlers = [];
+  }
+
+  /**
+   * Registers action handler
+   *
+   * @param {String} String pattenr
+   * @param {Function} Action Handler (reducer)
+   * @param {Matcher} [Optional] Specific Matcher implementation
+   */
+  case(pattern, actionHandler, matcherImpl) {
+    const matcher = matcherImpl ?
+      matcherImpl(pattern) :
+      this.defaultMatcherImpl(pattern);
+
+    this.matchersWithHandlers.push({
+      matcher,
+      actionHandler
+    });
+
+    return this;
+  }
+
+  /**
+   * Creates the Reducer
+   *
+   * @return {Function} plain old redux Reducer
+   */
+  toReducer() {
+    return (appState = this.initialAppState, action) => {
+      if (action) {
+        if (!action.type) {
+          // Logging is technically impurity yet this serves
+          // just for developer to eliminate any potential issues
+          // in development phase.
+          warn('It looks like you have provided invalid action. ' +
+               'Please make sure that you have dispatched an object with at least ' +
+               'one field (type) identifiying type of the dispatched action. Or ' +
+               'make sure that you are using appropriate middleware (eg. redux-thunk) ' +
+               'which translates non-standard actions to plain old Redux actions in ' +
+               'form of objects.');
```

Comment:
"It looks like you have provided an invalid action. Make sure that the dispatched action has a `type` property or use the appropriate middleware (e.g. redux-thunk) to translate non-standard actions into plain old Redux action objects."